src/SignalR/common/Http.Connections.Common/src/NegotiateProtocol.cs

```diff
@@ -247,5 +338,138 @@ private static AvailableTransport ParseAvailableTransport(JsonTextReader reader)
+                        } else if (memberName.SequenceEqual(TransferFormatsPropertyNameBytes))
+                        {
+                            CheckRead(ref reader);
+                            EnsureArrayStart(ref reader);
+
+                            var completed = false;
+
+                            availableTransport.TransferFormats = new List<string>();
+                            while (!completed && CheckRead(ref reader))
+                            {
+                                switch (reader.TokenType)
+                                {
+                                    case JsonTokenType.String:
+                                        availableTransport.TransferFormats.Add(reader.GetString());
+                                        break;
+                                    case JsonTokenType.EndArray:
+                                        completed = true;
+                                        break;
+                                    default:
+                                        throw new InvalidDataException($"Unexpected token '{reader.TokenType}' when reading transfer formats JSON.");
+                                }
+                            }
+                        }
+                        else
+                        {
+                            Skip(ref reader);
+                        }
+                        break;
+                    case JsonTokenType.EndObject:
+                        if (availableTransport.Transport == null)
+                        {
+                            throw new InvalidDataException($"Missing required property '{TransportPropertyName}'.");
+                        }
+
+                        if (availableTransport.TransferFormats == null)
+                        {
+                            throw new InvalidDataException($"Missing required property '{TransferFormatsPropertyName}'.");
+                        }
+
+                        return availableTransport;
+                    default:
+                        throw new InvalidDataException($"Unexpected token '{reader.TokenType}' when reading available transport JSON.");
+                }
+            }
+
+            throw new InvalidDataException("Unexpected end when reading JSON.");
+        }
+
+        private static bool CheckRead(ref Utf8JsonReader reader)
+        {
+            if (!reader.Read())
+            {
+                throw new InvalidDataException("Unexpected end when reading JSON.");
+            }
+
+            return true;
+        }
+
+        private static void EnsureObjectStart(ref Utf8JsonReader reader)
+        {
+            if (reader.TokenType != JsonTokenType.StartObject)
+            {
+                throw new InvalidDataException($"Unexpected JSON Token Type '{GetTokenString(reader.TokenType)}'. Expected a JSON Object.");
+            }
+        }
+
+        private static string GetTokenString(JsonTokenType tokenType)
```

Comment:
Why is this method necessary at all?
The returned special case strings have the same value as the enum names?